Disabled remote-debugging by default and appshell.app.getRemoteDebuggingPort() needs a callback argument to work

jha-g opened the pull request:
* Disabling remote debugging by default, and it can be enabled by passing a command line arg as --remote-debugging-port=xxxx
* Changed appshell.app.getRemoteDebuggingPort() implementation to require a callback
* Added validation for port range, and also without remote-debugging-port we will not be able to debug brackets using a browser, and if browser debugging is disabled, appshell.app.getRemoteDebuggingPort(callback) will print 0

shubhsnov commented on appshell/cefclient.cpp:
```diff
@@ -95,7 +98,19 @@ void AppGetSettings(CefSettings& settings, CefRefPtr<CefCommandLine> command_lin
+  CefString debugger_port = command_line->GetSwitchValue("remote-debugging-port");
+  if (!debugger_port.empty()) {
+    int port = atoi(debugger_port.ToString().c_str());
+    static const int max_port_num = static_cast<int>(std::numeric_limits<uint16_t>::max());
```
What happens if the port is already in use?

jha-g replied:
I checked with `python -m SimpleHTTPServer 9999` and Brackets snatches the port from Python HTTP Server.

shubhsnov commented on appshell/appshell_extensions.cpp:
```diff
@@ -842,6 +842,8 @@ class ProcessMessageDelegate : public ClientHandler::ProcessMessageDelegate {
             uberDict->SetList(0, dirContents);
             uberDict->SetList(1, allStats);
             responseArgs->SetList(2, uberDict);
+        } else if (message_name == "GetRemoteDebuggingPort") {
```
We're changing the normal function to one that takes callback? Does this have any impact on existing usages? This is a breaking change, though I guess no extensions will ideally use this.

jha-g replied:
Yes, it does impact one use case, and fixed that with commit: https://github.com/adobe/brackets/pull/14956/commits/8677756aa443b98523852f86bfcbfc1a77d93478

shubhsnov commented on appshell/cefclient.cpp:
```diff
@@ -95,7 +98,19 @@ void AppGetSettings(CefSettings& settings, CefRefPtr<CefCommandLine> command_lin
+  CefString debugger_port = command_line->GetSwitchValue("remote-debugging-port");
+  if (!debugger_port.empty()) {
+    int port = atoi(debugger_port.ToString().c_str());
+    static const int max_port_num = static_cast<int>(std::numeric_limits<uint16_t>::max());
```
Can we get away without needing limits header? 
I am not comfortable with the need for unsetting standard functions.
You can compute uint16 max as such, will work for all platforms
uint16 max_limit = 0;
max_limit = ~max_limit;
max_limit = max_limit >> 1

jha-g replied:
std::numeric_limits<> is C++ equivalent of C macros denoting MAX and MIN of a default numeric data types. So, instead of `std::numeric_limits<uint16_t>::max()`, I could use `UINT16_MAX` defined in [stdint.h](https://en.cppreference.com/w/cpp/header/cstdint)

jha-g replied:
Don't bother about header, it's C++, not JS.

jha-g commented:
Addressed all the review comments.

swmitra commented:
@vickramdhawal Can you please take a quick look?

vickramdhawal commented on appshell/cefclient.cpp:
```diff
@@ -95,7 +96,21 @@ void AppGetSettings(CefSettings& settings, CefRefPtr<CefCommandLine> command_lin
-  settings.remote_debugging_port = REMOTE_DEBUGGING_PORT;
+  CefString debugger_port = command_line->GetSwitchValue("remote-debugging-port");
+  if (!debugger_port.empty()) {
+    int port = atoi(debugger_port.ToString().c_str());
```
using atom seems acceptable in this case since the string is coming from the CommandLine handler. Better would be to use strtol.

jha-g replied:
Changed parsing to use `strtol`.

vickramdhawal commented on appshell/cefclient.cpp:
```diff
@@ -19,6 +19,7 @@
 #include "config.h"
 
 CefRefPtr<ClientHandler> g_handler;
+int g_remote_debugging_port = 0;
```
Since, we are going to use atom or strtol further, as suggested by you using int is better for comparisons.

jha-g commented on appshell/config.h:
```diff
@@ -82,7 +82,7 @@
 #endif
 
-#define REMOTE_DEBUGGING_PORT 9234
+extern int g_remote_debugging_port;
```
Done.

jha-g commented on appshell/cefclient.cpp:
```diff
@@ -95,7 +96,21 @@ void AppGetSettings(CefSettings& settings, CefRefPtr<CefCommandLine> command_lin
+  if (!debugger_port.empty()) {
+    int port = atoi(debugger_port.ToString().c_str());
+    static const int max_port_num = 65535;
+    static const int max_reserved_port_num = 1024;
```
Using `strtol` for integer parsing, so moved to `signed long`.

jha-g commented:
Addressed review comments. It is not possible to use `g_handler` to hold remote-debugging-port as `g_handler` does not get initialized while we are parsing command line.

vickramdhawal commented on appshell/cefclient.cpp:
```diff
@@ -95,7 +97,25 @@ void AppGetSettings(CefSettings& settings, CefRefPtr<CefCommandLine> command_lin
+        LOG(ERROR) << "Error while parsing remote-debugging-port arg: "<< debugger_port.ToString();
+        errno = 0;
+    }
+    static const long max_port_num = 65535;
```
all of this should then be in the else case if errno is not ERANGE, since we know the port was not parsed properly and comparisons further may not give desired results.

jha-g replied:
Done.
